src/tools/wasm-link.c

```diff
@@ -759,61 +563,130 @@ static WasmBool write_combined_section(Context* ctx,
   return WASM_TRUE;
 }
 
+typedef struct ExportInfo {
+  Export* export;
+  InputBinary* binary;
+} ExportInfo;
+WASM_DEFINE_VECTOR(export_info, ExportInfo);
+
+static void resolve_symbols(Context* ctx) {
+  /* Create hashmap of all exported symbols from all inputs */
+  WasmBindingHash export_map;
+  WASM_ZERO_MEMORY(export_map);
+  ExportInfoVector export_list;
+  WASM_ZERO_MEMORY(export_list);
+
+  size_t i, j;
+  for (i = 0; i < ctx->inputs.size; i++) {
+    InputBinary* binary = &ctx->inputs.data[i];
+    for (j = 0; j < binary->exports.size; j++) {
+      Export* export = &binary->exports.data[j];
+      ExportInfo* info = wasm_append_export_info(ctx->allocator, &export_list);
+      info->export = export;
+      info->binary = binary;
+      WasmStringSlice name = wasm_dup_string_slice(ctx->allocator, export->name);
+      WasmBinding* binding = wasm_insert_binding(ctx->allocator, &export_map, &name);
+      binding->index = export_list.size - 1;
+    }
+  }
+
+  /*
+   * Iterate through all imported functions resolving them against exported
+   * ones.
+   */
+  for (i = 0; i < ctx->inputs.size; i++) {
+    InputBinary* binary = &ctx->inputs.data[i];
+    for (j = 0; j < binary->function_imports.size; j++) {
+      FunctionImport* import = &binary->function_imports.data[j];
+      int export_index = wasm_find_binding_index_by_name(&export_map, &import->name);
+      if (export_index == -1) {
+        if (!s_relocatable)
+          WASM_FATAL("undefined symbol: " PRIstringslice "\n", WASM_PRINTF_STRING_SLICE_ARG(import->name));
+        continue;
+      }
+
+      /* We found the symbol exported by another module */
+      ExportInfo* export_info = &export_list.data[export_index];
+
+      /* TODO(sbc): verify the foriegn function has the correct signature */
+      import->active = WASM_FALSE;
+      import->foreign_binary = export_info->binary;
+      import->foreign_index = export_info->export->index;
+      binary->active_function_imports--;
+    }
+  }
+
+  wasm_destroy_export_info_vector(ctx->allocator, &export_list);
+  wasm_destroy_binding_hash(ctx->allocator, &export_map);
+}
+
+static void calculate_reloc_offsets(Context* ctx) {
+  size_t i, j;
+  uint32_t memory_page_offset = 0;
+  uint32_t type_count = 0;
+  uint32_t global_count = 0;
+  uint32_t function_count = 0;
+  uint32_t total_function_imports = 0;
+  uint32_t total_global_imports = 0;
+  for (i = 0; i < ctx->inputs.size; i++) {
+    InputBinary* binary = &ctx->inputs.data[i];
+    /* The imported_function_index_offset is the sum of all the function
+     * imports from objects that precede this one.  i.e. the current running
+     * total */
+    binary->imported_function_index_offset = total_function_imports;
+    binary->imported_global_index_offset = total_global_imports;
+    binary->memory_page_offset = memory_page_offset;
+    memory_page_offset += binary->memory_page_count;
```

Comment:
Currently we generate relocation for the ELEM section, but I don't have any way to tagging data which refers to addresses of functions.   Such references would generate relocation of type WASM_RELOC_TABLE_INDEX_SLEB/I32 which would then needs to be correctly handled.  So yes, we don't support that in wabt yet.